jupyter_server/files/handlers.py

```diff
@@ -38,41 +46,229 @@ async def get(self, path, include_body=True):
+
+    if archive_format == "zip":
+        archive_file = zipfile.ZipFile(
+            fileobj, mode="w", compression=zipfile.ZIP_DEFLATED
+        )
+        archive_file.add = archive_file.write
+    elif archive_format in ["tgz", "tar.gz"]:
+        archive_file = tarfile.open(fileobj=fileobj, mode="w|gz")
+    elif archive_format in ["tbz", "tbz2", "tar.bz", "tar.bz2"]:
+        archive_file = tarfile.open(fileobj=fileobj, mode="w|bz2")
+    elif archive_format in ["txz", "tar.xz"]:
+        archive_file = tarfile.open(fileobj=fileobj, mode="w|xz")
+    else:
+        raise ValueError("'{}' is not a valid archive format.".format(archive_format))
+    return archive_file
+
+
+def make_reader(archive_path):
+
+    archive_format = "".join(archive_path.suffixes)[1:]
+
+    if archive_format == "zip":
+        archive_file = zipfile.ZipFile(archive_path, mode="r")
+    elif archive_format in ["tgz", "tar.gz"]:
+        archive_file = tarfile.open(archive_path, mode="r|gz")
+    elif archive_format in ["tbz", "tbz2", "tar.bz", "tar.bz2"]:
+        archive_file = tarfile.open(archive_path, mode="r|bz2")
+    elif archive_format in ["txz", "tar.xz"]:
+        archive_file = tarfile.open(archive_path, mode="r|xz")
+    else:
+        raise ValueError("'{}' is not a valid archive format.".format(archive_format))
+    return archive_file
+
+
+class DirectoryHandler(JupyterHandler):
+    @web.authenticated
+    async def get(self, archive_path, include_body=False):
+
+        # /directories/ requests must originate from the same site
+        self.check_xsrf_cookie()
+        cm = self.contents_manager
+
+        if cm.is_hidden(archive_path) and not cm.allow_hidden:
+            self.log.info("Refusing to serve hidden file, via 404 Error")
+            raise web.HTTPError(404)
+
+        archive_token = self.get_argument("archiveToken")
+        archive_format = self.get_argument("archiveFormat", "zip")
+        if archive_format not in SUPPORTED_FORMAT:
+            self.log.error("Unsupported format {}.".format(archive_format))
+            raise web.HTTPError(404)
+
+        root_dir = pathlib.Path(cm.root_dir)
+        archive_path = root_dir / url2path(archive_path)
+        archive_filename = archive_path.with_suffix(".{}".format(archive_format)).name
+
+        self.log.info(
+            "Prepare {} for archiving and downloading.".format(archive_filename)
+        )
+        self.set_header("content-type", "application/octet-stream")
+        self.set_header("cache-control", "no-cache")
+        self.set_header(
+            "content-disposition", "attachment; filename={}".format(archive_filename)
+        )
+
+        self.canceled = False
+
+        # The original extension is using `ioloop.PeriodicCallback` and `ioloop.IOLoop.current().run_in_executor`
+        # but it does not seem to work here (500 internal HTTP error).
+        # This trategy is nice when compressing very large directory. Any idea how to integrate that is welcome.
```

Comment:
I agree we'll want a non-blocking archive (and extract).  Could you please post the complete traceback associated with the 500 error?

Reply:
See my last commit. Associated pytest logs:

```bash
$ pytest --verbosity 3 tests/test_files.py 
/home/hadim/local/conda/envs/jupyter/lib/python3.8/site-packages/_pytest/compat.py:333: PytestDeprecationWarning: The TerminalReporter.writer attribute is deprecated, use TerminalReporter._tw instead at your own risk.
See https://docs.pytest.org/en/latest/deprecations.html#terminalreporter-writer for more information.
  return getattr(object, name, default)
================================================================================================================================= test session starts =================================================================================================================================
platform linux -- Python 3.8.3, pytest-5.4.3, py-1.9.0, pluggy-0.13.1 -- /home/hadim/local/conda/envs/jupyter/bin/python
cachedir: .pytest_cache
rootdir: /home/hadim/Drive/Documents/Code/Misc/jupyter/jupyter_server, inifile: pytest.ini
plugins: xonsh-0.9.18, tornasync-0.6.0.post2, jupyter-server-0.3.0
collected 6 items                                                                                                                                                                                                                                                                     

tests/test_files.py::test_hidden_files PASSED                                                                                                                                                                                                                                   [ 16%]
tests/test_files.py::test_contents_manager PASSED                                                                                                                                                                                                                               [ 33%]
tests/test_files.py::test_download PASSED                                                                                                                                                                                                                                       [ 50%]
tests/test_files.py::test_old_files_redirect PASSED                                                                                                                                                                                                                             [ 66%]
tests/test_files.py::test_download_directory FAILED                                                                                                                                                                                                                             [ 83%]
tests/test_files.py::test_extract_directory PASSED                                                                                                                                                                                                                              [100%]

====================================================================================================================================== FAILURES =======================================================================================================================================
_______________________________________________________________________________________________________________________________ test_download_directory _______________________________________________________________________________________________________________________________

fetch = <function fetch.<locals>.client_fetch at 0x7fb44ad1edc0>, serverapp = <jupyter_server.serverapp.ServerApp object at 0x7fb44acc4d00>, root_dir = PosixPath('/tmp/pytest-of-hadim/pytest-4/test_download_directory0/root_dir')

    async def test_download_directory(fetch, serverapp, root_dir):
        # Create a dummy directory.
        root_dir = pathlib.Path(str(root_dir))
        archive_dir_path = root_dir / "download-archive-dir"
        archive_dir_path.mkdir()
        with open(archive_dir_path / "test1.txt", "w") as f:
            f.write("hello1")
        with open(archive_dir_path / "test2.txt", "w") as f:
            f.write("hello2")
        with open(archive_dir_path / "test3.md", "w") as f:
            f.write("hello3")
    
        # Try to download the created folder.
        file_lists = {
            "download-archive-dir/test2.txt",
            "download-archive-dir/test1.txt",
            "download-archive-dir/test3.md",
        }
    
        archive_formats = {
            "zip": "r",
            "tgz": "r|gz",
            "tar.gz": "r|gz",
            "tbz": "r|bz2",
            "tbz2": "r|bz2",
            "tar.bz": "r|bz2",
            "tar.bz2": "r|bz2",
            "txz": "r|xz",
            "tar.xz": "r|xz",
        }
    
        archive_token = 59487596
    
        for archive_format, mode in archive_formats.items():
            params = dict(archiveToken=archive_token, archiveFormat=archive_format)
            dir_path = str(archive_dir_path.relative_to(root_dir))
>           r = await fetch("directories", dir_path, method="GET", params=params)
E           tornado.httpclient.HTTPClientError: HTTP 500: Internal Server Error

tests/test_files.py:179: HTTPClientError
-------------------------------------------------------------------------------------------------------------------------------- Captured stderr setup --------------------------------------------------------------------------------------------------------------------------------
[D 2020-07-07 07:59:50.846 ServerApp] Config changed:
[D 2020-07-07 07:59:50.847 ServerApp] {'NotebookNotary': {'db_file': ':memory:'}}
--------------------------------------------------------------------------------------------------------------------------------- Captured log setup ----------------------------------------------------------------------------------------------------------------------------------
DEBUG    ServerApp:application.py:179 Searching ['/home/hadim/Drive/Documents/Code/Misc/jupyter/jupyter_server', '/tmp/pytest-of-hadim/pytest-4/test_download_directory0/config', '/tmp/pytest-of-hadim/pytest-4/test_download_directory0/env/etc/jupyter', '/tmp/pytest-of-hadim/pytest-4/test_download_directory0/etc/jupyter'] for config files
DEBUG    ServerApp:application.py:556 Looking for jupyter_config in /tmp/pytest-of-hadim/pytest-4/test_download_directory0/etc/jupyter
DEBUG    ServerApp:application.py:556 Looking for jupyter_config in /tmp/pytest-of-hadim/pytest-4/test_download_directory0/env/etc/jupyter
DEBUG    ServerApp:application.py:556 Looking for jupyter_config in /tmp/pytest-of-hadim/pytest-4/test_download_directory0/config
DEBUG    ServerApp:application.py:556 Looking for jupyter_config in /home/hadim/Drive/Documents/Code/Misc/jupyter/jupyter_server
DEBUG    ServerApp:application.py:556 Looking for jupyter_server_config in /tmp/pytest-of-hadim/pytest-4/test_download_directory0/etc/jupyter
DEBUG    ServerApp:application.py:556 Looking for jupyter_server_config in /tmp/pytest-of-hadim/pytest-4/test_download_directory0/env/etc/jupyter
DEBUG    ServerApp:application.py:556 Looking for jupyter_server_config in /tmp/pytest-of-hadim/pytest-4/test_download_directory0/config
DEBUG    ServerApp:application.py:556 Looking for jupyter_server_config in /home/hadim/Drive/Documents/Code/Misc/jupyter/jupyter_server
DEBUG    ServerApp:config_manager.py:98 Paths used for configuration of jupyter_server_config: 
        /tmp/pytest-of-hadim/pytest-4/test_download_directory0/config/jupyter_server_config.json
DEBUG    ServerApp:config_manager.py:98 Paths used for configuration of jupyter_server_config: 
        /tmp/pytest-of-hadim/pytest-4/test_download_directory0/env/etc/jupyter/jupyter_server_config.json
DEBUG    ServerApp:config_manager.py:98 Paths used for configuration of jupyter_server_config: 
        /tmp/pytest-of-hadim/pytest-4/test_download_directory0/etc/jupyter/jupyter_server_config.json
INFO     ServerApp:serverapp.py:1936 Serving notebooks from local directory: /tmp/pytest-of-hadim/pytest-4/test_download_directory0/root_dir
INFO     ServerApp:serverapp.py:1936 Jupyter Server 0.3.0 is running at:
INFO     ServerApp:serverapp.py:1936 http://localhost:42833/?token=...
INFO     ServerApp:serverapp.py:1936  or http://127.0.0.1:42833/?token=...
INFO     ServerApp:serverapp.py:1937 Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).
---------------------------------------------------------------------------------------------------------------------------------- Captured log call ----------------------------------------------------------------------------------------------------------------------------------
DEBUG    ServerApp:login.py:205 Accepting token-authenticated connection from 127.0.0.1
ERROR    ServerApp:web.py:1788 Uncaught exception GET /directories/download-archive-dir?archiveToken=59487596&archiveFormat=zip (127.0.0.1)
HTTPServerRequest(protocol='http', host='127.0.0.1:42833', method='GET', uri='/directories/download-archive-dir?archiveToken=59487596&archiveFormat=zip', version='HTTP/1.1', remote_ip='127.0.0.1')
Traceback (most recent call last):
  File "/home/hadim/local/conda/envs/jupyter/lib/python3.8/site-packages/tornado/web.py", line 1703, in _execute
    result = await result
TypeError: object async_generator can't be used in 'await' expression
DEBUG    ServerApp:handlers.py:241 Using contents: services/contents
DEBUG    ServerApp:handlers.py:241 Using contents: services/contents
ERROR    ServerApp:log.py:48 {
  "Authorization": "token c4441194",
  "Connection": "close",
  "Host": "127.0.0.1:42833",
  "Accept-Encoding": "gzip"
}
ERROR    ServerApp:log.py:49 500 GET /directories/download-archive-dir?archiveToken=59487596&archiveFormat=zip (127.0.0.1) 5.95ms referer=None
-------------------------------------------------------------------------------------------------------------------------------- Captured log teardown --------------------------------------------------------------------------------------------------------------------------------
INFO     ServerApp:serverapp.py:1808 Shutting down 0 kernels
=============================================================================================================================== short test summary info ===============================================================================================================================
FAILED tests/test_files.py::test_download_directory - tornado.httpclient.HTTPClientError: HTTP 500: Internal Server Error
============================================================================================================================= 1 failed, 5 passed in 0.30s =============================================================================================================================
```

Reply:
It looks like my last commit fixes this. I had to replace `yield` by `await`. I guess it's because before I was using `@gen.coroutine` and now I use `async`.